x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -825,4 +839,115 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
         plan.forEachDown(p -> FeatureMetric.set(p, b));
         return b;
     }
+
+    private static class ImplicitCasting extends ParameterizedRule<LogicalPlan, LogicalPlan, AnalyzerContext> {
```

Comment:
Modify the AnalyzerContext to return an EsqlFunctionRegistry or cast the FunctionRegistry to EsqlFunctionRegistry inside the rule.